File: shim-python/enterprise.py
from __future__ import annotations
import hashlib, html, json, os, re, secrets

PUBLIC_BASE=os.getenv('FVS_PUBLIC_BASE_URL','http://localhost:8080').rstrip('/')

def agent_ok(env):
    supplied=env.get('HTTP_X_AGENT_KEY','')
    expected_hash=os.getenv('FVS_AGENT_KEY_SHA256','').strip().lower()
    if expected_hash:
        return secrets.compare_digest(hashlib.sha256(supplied.encode('utf-8')).hexdigest(),expected_hash)
    if os.getenv('FVS_ENV','development')=='production': return False
    expected=os.getenv('FVS_AGENT_KEY','')
    return bool(expected and secrets.compare_digest(supplied,expected))

def agent_id(env):
    raw=(env.get('HTTP_X_AGENT_ID') or '').strip()
    return raw[:64] if re.fullmatch(r'[A-Za-z0-9._:@-]{1,64}',raw) else ''

def seo_html(page):
    e=lambda v:html.escape(str(v or ''),quote=True)
    prop=page.get('property') or {}
    canonical=PUBLIC_BASE+str(page.get('canonical_path') or '')
    image=str(page.get('og_image_url') or prop.get('image_url') or '')
    faq=page.get('faq') if isinstance(page.get('faq'),list) else []
    page_type=str(page.get('page_type') or 'property')

    if page_type == 'property':
        accommodation={
            '@type':'Accommodation',
            'name':prop.get('unit_name') or page.get('h1'),
            'identifier':prop.get('slot_id') or None,
        }
        if prop.get('max_guests'):
            accommodation['occupancy']={'@type':'QuantitativeValue','maxValue':int(prop.get('max_guests') or 0)}
        ld={
            '@context':'https://schema.org',
            '@type':'LodgingBusiness',
            'name':prop.get('resort') or page.get('h1'),
            'description':page.get('meta_description'),
            'url':canonical,
            'containsPlace':accommodation,
        }
        if image: ld['image']=[image]
        if prop.get('city') or prop.get('country'):
            ld['address']={'@type':'PostalAddress','addressLocality':prop.get('city') or '','addressCountry':prop.get('country') or ''}
        if prop.get('price_minor') is not None and prop.get('currency'):
            ld['offers']={
                '@type':'Offer',
                'price':f"{int(prop.get('price_minor') or 0)/100:.2f}",
                'priceCurrency':str(prop.get('currency')),
                'availability':'https://schema.org/InStock',
                'url':canonical,
                'itemOffered':accommodation,
            }
    else:
        schema_type='CollectionPage' if page_type in ('collection','destination') else 'WebPage'
        ld={
            '@context':'https://schema.org','@type':schema_type,
            'name':page.get('h1') or page.get('title'),
            'description':page.get('meta_description'),'url':canonical,
        }
        if image: ld['primaryImageOfPage']={'@type':'ImageObject','url':image}

    breadcrumb={'@context':'https://schema.org','@type':'BreadcrumbList','itemListElement':[{'@type':'ListItem','position':1,'name':'FVS','item':PUBLIC_BASE+'/'},{'@type':'ListItem','position':2,'name':page.get('h1'),'item':canonical}]}
    jsonlds=''.join('<script type="application/ld+json">'+json.dumps(x,ensure_ascii=False,separators=(',',':')).replace('</','<\\/')+'</script>' for x in (ld,breadcrumb))
    faq_html=''.join('<details><summary>'+e(x.get('question'))+'</summary><p>'+e(x.get('answer'))+'</p></details>' for x in faq if isinstance(x,dict) and x.get('question') and x.get('answer'))
    og_image='<meta property="og:image" content="'+e(image)+'">' if image else ''
    hero='<img class="seo-hero" src="'+e(image)+'" alt="'+e(page.get('h1'))+'" loading="eager" fetchpriority="high">' if image else ''
    faq_section='<section class="seo-faq"><h2>Preguntas frecuentes</h2>'+faq_html+'</section>' if faq_html else ''
    price=''
    if prop.get('price_minor') is not None and prop.get('currency'):
        price=f"{int(prop.get('price_minor') or 0)/100:.2f} {prop.get('currency') or ''}"
    robots='index,follow,max-image-preview:large' if page.get('indexable',True) else 'noindex,follow'
    property_facts=''
    property_cta=''
    eyebrow=e(prop.get('city') or prop.get('country'))
    if page_type == 'property':
        property_facts=(
            '<dl class="seo-facts"><div><dt>Entrada</dt><dd>'+e(prop.get('check_in'))+'</dd></div>'
            '<div><dt>Salida</dt><dd>'+e(prop.get('check_out'))+'</dd></div>'
            '<div><dt>Huéspedes</dt><dd>'+e(prop.get('max_guests'))+'</dd></div>'
            '<div><dt>Precio</dt><dd>'+e(price)+'</dd></div></dl>'
        )
        property_cta='<p><a class="primary seo-cta" href="/?slot='+e(prop.get('slot_id'))+'&utm_source=organic&utm_medium=seo&utm_campaign=property_page">Ver disponibilidad y reservar</a></p>'
    return (
        '<!doctype html><html lang="'+e(page.get('locale') or 'es-MX')+'"><head><meta charset="utf-8">'
        '<meta name="viewport" content="width=device-width,initial-scale=1"><title>'+e(page.get('title'))+'</title>'
        '<meta name="description" content="'+e(page.get('meta_description'))+'"><meta name="robots" content="'+robots+'">'
        '<link rel="canonical" href="'+e(canonical)+'"><meta property="og:type" content="website">'
        '<meta property="og:title" content="'+e(page.get('title'))+'"><meta property="og:description" content="'+e(page.get('meta_description'))+'">'
        '<meta property="og:url" content="'+e(canonical)+'">'+og_image+'<meta name="twitter:card" content="summary_large_image">'
        '<link rel="stylesheet" href="/css/base.css">'+jsonlds+'</head><body>'
        '<main class="shell"><header class="topbar"><a class="brand" href="/">FVS<small>Vacation Store</small></a></header>'
        '<article class="seo-page"><p class="eyebrow">'+eyebrow+'</p><h1>'+e(page.get('h1'))+'</h1>'+hero+
        '<p class="seo-lead">'+e(page.get('body_text') or page.get('meta_description'))+'</p>'+property_facts+property_cta+faq_section+
        '</article></main></body></html>'
    )

def sitemap_xml(pages):
    items=[]
    for p in pages:
        loc=PUBLIC_BASE+str(p.get('canonical_path') or '')
        items.append('<url><loc>'+html.escape(loc)+'</loc><lastmod>'+html.escape(str(p.get('updated_at') or '')[:10])+'</lastmod><changefreq>'+html.escape(str(p.get('changefreq') or 'weekly'))+'</changefreq><priority>'+f"{float(p.get('priority') or .7):.2f}"+'</priority></url>')
    return '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">'+''.join(items)+'</urlset>'

File: shim-php/Enterprise.php
<?php declare(strict_types=1);

function agentOk(): bool {
    $sup = $_SERVER['HTTP_X_AGENT_KEY'] ?? '';
    $hash = strtolower(trim((string)(getenv('FVS_AGENT_KEY_SHA256') ?: '')));
    if ($hash !== '') return hash_equals($hash, hash('sha256', $sup));
    if ((getenv('FVS_ENV') ?: 'development') === 'production') return false;
    $expected = getenv('FVS_AGENT_KEY') ?: '';
    return $expected !== '' && hash_equals($expected, $sup);
}

function agentId(): string {
    $v = trim((string)($_SERVER['HTTP_X_AGENT_ID'] ?? ''));
    return preg_match('/^[A-Za-z0-9._:@-]{1,64}$/D', $v) ? $v : '';
}

function supportRespond(int $code, array $data, string $secret): never {
    http_response_code($code);
    header('Content-Type: application/json; charset=utf-8');
    header('Cache-Control: no-store');
    setcookie('fvs_support_secret', $secret, [
        'expires' => time() + 604800,
        'path' => '/api/v1/support',
        'secure' => getenv('FVS_COOKIE_SECURE') === '1',
        'httponly' => true,
        'samesite' => 'Lax',
    ]);
    echo json_encode($data, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
    exit;
}

function h(mixed $v): string {
    return htmlspecialchars((string)($v ?? ''), ENT_QUOTES | ENT_SUBSTITUTE, 'UTF-8');
}

function seoHtml(array $p): string {
    $base = rtrim(getenv('FVS_PUBLIC_BASE_URL') ?: 'http://localhost:8080', '/');
    $prop = $p['property'] ?? [];
    $canonical = $base . (string)($p['canonical_path'] ?? '');
    $image = (string)($p['og_image_url'] ?? $prop['image_url'] ?? '');
    $pageType = (string)($p['page_type'] ?? 'property');
    $faq = is_array($p['faq'] ?? null) ? $p['faq'] : [];

    if ($pageType === 'property') {
        $accommodation = [
            '@type' => 'Accommodation',
            'name' => $prop['unit_name'] ?? $p['h1'] ?? '',
        ];
        if (!empty($prop['slot_id'])) $accommodation['identifier'] = $prop['slot_id'];
        if (!empty($prop['max_guests'])) {
            $accommodation['occupancy'] = ['@type' => 'QuantitativeValue', 'maxValue' => (int)$prop['max_guests']];
        }
        $ld = [
            '@context' => 'https://schema.org',
            '@type' => 'LodgingBusiness',
            'name' => $prop['resort'] ?? $p['h1'] ?? '',
            'description' => $p['meta_description'] ?? '',
            'url' => $canonical,
            'containsPlace' => $accommodation,
        ];
        if ($image !== '') $ld['image'] = [$image];
        if (($prop['city'] ?? '') !== '' || ($prop['country'] ?? '') !== '') {
            $ld['address'] = ['@type' => 'PostalAddress', 'addressLocality' => $prop['city'] ?? '', 'addressCountry' => $prop['country'] ?? ''];
        }
        if (array_key_exists('price_minor', $prop) && !empty($prop['currency'])) {
            $ld['offers'] = [
                '@type' => 'Offer',
                'price' => number_format(((int)$prop['price_minor']) / 100, 2, '.', ''),
                'priceCurrency' => $prop['currency'],
                'availability' => 'https://schema.org/InStock',
                'url' => $canonical,
                'itemOffered' => $accommodation,
            ];
        }
    } else {
        $ld = [
            '@context' => 'https://schema.org',
            '@type' => in_array($pageType, ['collection', 'destination'], true) ? 'CollectionPage' : 'WebPage',
            'name' => $p['h1'] ?? $p['title'] ?? '',
            'description' => $p['meta_description'] ?? '',
            'url' => $canonical,
        ];
        if ($image !== '') $ld['primaryImageOfPage'] = ['@type' => 'ImageObject', 'url' => $image];
    }

    $bc = [
        '@context' => 'https://schema.org', '@type' => 'BreadcrumbList',
        'itemListElement' => [
            ['@type' => 'ListItem', 'position' => 1, 'name' => 'FVS', 'item' => $base . '/'],
            ['@type' => 'ListItem', 'position' => 2, 'name' => $p['h1'] ?? '', 'item' => $canonical],
        ],
    ];
    $jsonld = '';
    foreach ([$ld, $bc] as $obj) {
        $encoded = json_encode($obj, JSON_UNESCAPED_UNICODE | JSON_UNESCAPED_SLASHES);
        $jsonld .= '<script type="application/ld+json">' . str_replace('</', '<\/', (string)$encoded) . '</script>';
    }
    $faqHtml = '';
    foreach ($faq as $x) {
        if (is_array($x) && !empty($x['question']) && !empty($x['answer'])) {
            $faqHtml .= '<details><summary>' . h($x['question']) . '</summary><p>' . h($x['answer']) . '</p></details>';
        }
    }
    $price = '';
    if (array_key_exists('price_minor', $prop) && !empty($prop['currency'])) {
        $price = number_format(((int)$prop['price_minor']) / 100, 2, '.', '') . ' ' . h($prop['currency']);
    }
    $robots = !array_key_exists('indexable', $p) || $p['indexable'] ? 'index,follow,max-image-preview:large' : 'noindex,follow';
    $facts = '';
    $cta = '';
    if ($pageType === 'property') {
        $facts = '<dl class="seo-facts"><div><dt>Entrada</dt><dd>' . h($prop['check_in'] ?? '') . '</dd></div>'
            . '<div><dt>Salida</dt><dd>' . h($prop['check_out'] ?? '') . '</dd></div>'
            . '<div><dt>Huéspedes</dt><dd>' . h($prop['max_guests'] ?? '') . '</dd></div>'
            . '<div><dt>Precio</dt><dd>' . $price . '</dd></div></dl>';
        $cta = '<p><a class="primary seo-cta" href="/?slot=' . h($prop['slot_id'] ?? '') . '&amp;utm_source=organic&amp;utm_medium=seo&amp;utm_campaign=property_page">Ver disponibilidad y reservar</a></p>';
    }
    $hero = $image !== '' ? '<img class="seo-hero" src="' . h($image) . '" alt="' . h($p['h1'] ?? '') . '" loading="eager" fetchpriority="high">' : '';
    $faqSection = $faqHtml !== '' ? '<section class="seo-faq"><h2>Preguntas frecuentes</h2>' . $faqHtml . '</section>' : '';

    return '<!doctype html><html lang="' . h($p['locale'] ?? 'es-MX') . '"><head><meta charset="utf-8">'
        . '<meta name="viewport" content="width=device-width,initial-scale=1"><title>' . h($p['title'] ?? '') . '</title>'
        . '<meta name="description" content="' . h($p['meta_description'] ?? '') . '"><meta name="robots" content="' . h($robots) . '">'
        . '<link rel="canonical" href="' . h($canonical) . '"><meta property="og:type" content="website"><meta property="og:title" content="' . h($p['title'] ?? '') . '">'
        . '<meta property="og:description" content="' . h($p['meta_description'] ?? '') . '"><meta property="og:url" content="' . h($canonical) . '">'
        . ($image !== '' ? '<meta property="og:image" content="' . h($image) . '">' : '')
        . '<meta name="twitter:card" content="summary_large_image"><link rel="stylesheet" href="/css/base.css">' . $jsonld . '</head><body>'
        . '<main class="shell"><header class="topbar"><a class="brand" href="/">FVS<small>Vacation Store</small></a></header>'
        . '<article class="seo-page"><p class="eyebrow">' . h($prop['city'] ?? $prop['country'] ?? '') . '</p><h1>' . h($p['h1'] ?? '') . '</h1>' . $hero
        . '<p class="seo-lead">' . h($p['body_text'] ?? $p['meta_description'] ?? '') . '</p>' . $facts . $cta . $faqSection
        . '</article></main></body></html>';
}

function sitemapXml(array $pages): string {
    $base = rtrim(getenv('FVS_PUBLIC_BASE_URL') ?: 'http://localhost:8080', '/');
    $out = '';
    foreach ($pages as $p) {
        $out .= '<url><loc>' . h($base . ($p['canonical_path'] ?? '')) . '</loc><lastmod>' . h(substr((string)($p['updated_at'] ?? ''), 0, 10)) . '</lastmod>'
            . '<changefreq>' . h($p['changefreq'] ?? 'weekly') . '</changefreq><priority>' . number_format((float)($p['priority'] ?? .7), 2, '.', '') . '</priority></url>';
    }
    return '<?xml version="1.0" encoding="UTF-8"?><urlset xmlns="http://www.sitemaps.org/schemas/sitemap/0.9">' . $out . '</urlset>';
}
